app/controllers/msb/WhichCurrenciesController.scala

```diff
@@ -50,21 +50,31 @@ class WhichCurrenciesController @Inject() (val authConnector: AuthConnector,
     implicit authContext => implicit request => {
       Form2[WhichCurrencies](request.body) match {
         case f: InvalidForm =>
-          Future.successful(BadRequest(views.html.msb.which_currencies(f, edit)))
+          Future.successful(BadRequest(views.html.msb.which_currencies(removeEmptyFields(f), edit)))
         case ValidForm(_, data: WhichCurrencies) =>
               for {
                 msb <- dataCacheConnector.fetch[MoneyServiceBusiness](MoneyServiceBusiness.key)
                 _ <- dataCacheConnector.save[MoneyServiceBusiness](MoneyServiceBusiness.key,
                   updateCurrencies(msb, data))
-              } yield
-                  edit match {
-                    case true => Redirect(routes.SummaryController.get())
-                    case _ => Redirect(routes.UsesForeignCurrenciesController.get())
-                  }
+              } yield edit match {
+                case true => Redirect(routes.SummaryController.get())
+                case _ => Redirect(routes.UsesForeignCurrenciesController.get())
+              }
       }
     }
   }
 
+  // Validation removes empty fields. This removes them from the form data, so the errors align correctly with the fields.
+  def removeEmptyFields(f: InvalidForm): InvalidForm = {
+    val csrfToken = f.data.head
+    val fieldsWithData:Map[String, Seq[String]] = f.data
+      .filter(field => field._1.contains("currencies"))
+      .filter(field => field._2.exists(s => s.nonEmpty))
+      .zipWithIndex
+      .map((tuple: ((String, Seq[String]), Int)) => (tuple._1._1.replaceFirst("[\\d]", s"${tuple._2}"), tuple._1._2))
+    f.copy(data = fieldsWithData + csrfToken)
+  }
+
```

Comment:
Similar function exists in other controller, is there a way to generalise it and reuse? Can we have better naming here and split it if possible as it's hard to read?

Reply:
I'll look into this - they differ in the .map() as the forms are slightly different in structure but it'll be do-able.